### ISSUE TITLE
Add angle and signed angle?

### ISSUE THREAD

**Issue by quakeulf** [Unity-Technologies/Unity.Mathematics#84]
The greetings.

I have forked this master-branch and from this master-branch I have edited math.cs of which you can see here:

https://github.com/quakeulf/Unity.Mathematics/blob/master/src/Unity.Mathematics/math.cs

Specificially from line 98 to line 111 are my additions. I am also including them below this text so you can see in this field of text what I have proposed to add, but cannot, since I cannot seem to figure out how to do a request for pull:
        
```
        /// <summary>Returns the angle in degrees from 0 to 180 between two float3s.</summary>
        public static float angle(float3 from, float3 to)
        {
            return math.degrees(math.acos(math.dot(math.normalize(from), math.normalize(to))));
        }
        
        /// <summary>Returns the signed angle in degrees from 180 to -180 between two float3s.</summary>
        public static float anglesigned(float3 from, float3 to)
        {
            float angle = math.acos(math.dot(math.normalize(from), math.normalize(to)));
            float3 cross = math.cross(from, to);
            angle *= math.sign(math.dot(math.up(), cross));
            return math.degrees(angle);
        }
```
If they are already in there, let me know where. Preferably through documentation on your webcite.

**Comment by glabute** [comment]
This implementation is very noisy for small angles.  There are more precise (if somewhat slower) implementations.  Consider implementing 2 versions: fastAngle, and angle

**Comment by davejrodriguez** [comment]
The anglesigned implementation assumes math.up for the axis of rotation, which it will not be in every case. Anglesigned should take a third parameter of axis to replace math.up. See https://docs.unity3d.com/ScriptReference/Vector3.SignedAngle.html

**Comment by davejrodriguez** [comment]
The suggested anglesigned implementation assumes math.up for the axis of rotation, which it will not be in every case. Anglesigned should take a third parameter of axis to replace math.up. See https://docs.unity3d.com/ScriptReference/Vector3.SignedAngle.html

```public static float anglesigned(float3 from, float3 to, float3 axis)
    {
        float angle = math.acos(math.dot(math.normalize(from), math.normalize(to)));
        float sign = math.sign(math.dot(axis, math.cross(from, to)));
        angle *= sign;
        return math.degrees(angle);
    }
```

**Comment by davejrodriguez** [comment]
The suggested anglesigned implementation assumes math.up for the axis of rotation, which it will not be in every case. Anglesigned should take a third parameter of axis to replace math.up. See https://docs.unity3d.com/ScriptReference/Vector3.SignedAngle.html

```
/// <summary>Returns the signed angle in degrees from 180 to -180 between two float3s.</summary>
public static float anglesigned(float3 from, float3 to, float3 axis)
    {
        float angle = math.acos(math.dot(math.normalize(from), math.normalize(to)));
        float sign = math.sign(math.dot(axis, math.cross(from, to)));
        angle *= sign;
        return math.degrees(angle);
    }
```

**Comment by davejrodriguez** [comment]
The suggested anglesigned implementation assumes math.up for the axis of rotation, which it will not be in every case. Anglesigned should take a third parameter of axis to replace math.up. See https://docs.unity3d.com/ScriptReference/Vector3.SignedAngle.html

```
/// <summary>Returns the signed angle in degrees from 180 to -180 between two float3s.</summary>
public static float anglesigned(float3 from, float3 to, float3 axis)
{
    float angle = math.acos(math.dot(math.normalize(from), math.normalize(to)));
    float sign = math.sign(math.dot(axis, math.cross(from, to)));
    return math.degrees(angle * sign);
}
```

**Comment by Koopa1018** [comment]
> This implementation is very noisy for small angles. There are more precise (if somewhat slower) implementations.

Please do elaborate. I'm in need of an angle function, and I'm a total ignoramus regarding the requisite calculations.

**Comment by glabute** [comment]
Look here: https://forum.kerbalspaceprogram.com/index.php?/topic/164418-vector3angle-more-accurate-and-numerically-stable-at-small-angles-version/

**Comment by Koopa1018** [comment]
Thanks!

**Comment by nxrighthere** [comment]
@glabute Quite interesting stuff. Any other improvements/suggestions for Unity's standard math in terms of accuracy?

**Comment by quakeulf** [comment]
> Look here: https://forum.kerbalspaceprogram.com/index.php?/topic/164418-vector3angle-more-accurate-and-numerically-stable-at-small-angles-version/

This requires four  calls to magnitude, which incurs the square root calculation penalty, see:

https://docs.unity3d.com/ScriptReference/Vector3-magnitude.html

**Comment by Koopa1018** [comment]
Wonder how feasible it would be to include both versions....
And how standards-fitting.

**Comment by glabute** [comment]
Yes, the more precise version is a bit heavier.  We should have both versions IMO, clearly documented with the pros and cons.

**Comment by Indiefreaks** [comment]
I actually happen to be facing the lack of this feature in maths and used the actual implementation of Angle and SignedAngle found in the Unity.Physics Math class. Maybe Unity should just consider moving the code to Unity.Mathematics?

**Comment by Koopa1018** [comment]
_Ahhhhh._ I knew there was an official implementation somewhere...I just couldn't remember where :)

**Comment by Arugin** [comment]
Yes, these functions are very helpful. As well as double 3 equivalents.

**Comment by Koopa1018** [comment]
And 2D equivalents (that's my use case)!

**Comment by Aides359** [comment]
@Koopa1018 I can't seem to find the Unity.Physics.Math functions. Assuming this is the right class https://docs.unity3d.com/Packages/com.unity.physics@0.6/manual/index.html, there doesn't seem to be any function directly calculating the angle, neither 2D nor 3D.

**Comment by Koopa1018** [comment]
In all honesty, I never bothered to actually seek out the purported
Unity.Physics implementation.

I wonder if the Megacity demo might have it, though?

**Comment by ExodusOTH** [comment]
I'd love the inclusion of this whenever possible :) I've run into this need myself, and couldn't find the aforementioned Angle functions in Unity.Physics either. Currently I am going to try leveraging the example provided by _velenrendlich_ on [this forum post](https://forum.unity.com/threads/how-to-calculate-the-degrees-difference-y-axis-to-make-an-entity-look-at-a-target.1033579/#post-6693268)

`csharp
            public static float GetAngle(float3 a, float3 b)
            {
                return degrees(GetRad(a, b));
            }
     
            public static float GetRad(float3 a, float3 b)
            {
                return acos(clamp(dot(a, b) / (length(a) * length(b)), -1f, 1f));
            }

`

**Comment by VergilUa** [comment]
+1, 
There should be an equivalent for Vector3.Angle;
